prusti-specs/src/export_spec_rewriter.rs

```diff
@@ -0,0 +1,334 @@
+            path,
+            quote!(
+                $crate::#type_path_macro_ident!();
+                #tokens
+            ),
+            uses_same_level,
+        ),
+    }
+}
+
+pub fn rewrite_export_spec(path: &syn::Path, item: &syn::Item) -> syn::Result<TokenStream> {
+    let mut macros_tokens = TokenStream::new();
+    let mod_path = syn::Path {
+        leading_colon: None,
+        segments: syn::punctuated::Punctuated::new(),
+    };
+    match item {
+        syn::Item::Fn(item_fn) => {
+            rewrite_fn(
+                path,
+                mod_path,
+                item_fn,
+                &mut macros_tokens,
+                &TokenStream::new(),
+            );
+        }
+        syn::Item::Mod(item_mod) => {
+            rewrite_mod(
+                path,
+                mod_path,
+                item_mod,
+                &mut macros_tokens,
+                &TokenStream::new(),
+            )?;
+        }
+        syn::Item::Impl(item_impl) => {
+            rewrite_impl(
+                path,
+                mod_path,
+                item_impl,
+                &mut macros_tokens,
+                &TokenStream::new(),
+            )?;
+        }
+        syn::Item::Struct(item_struct) => {
+            rewrite_struct(
+                item_struct,
+                mod_path,
+                &mut macros_tokens,
+                &TokenStream::new(),
+            );
+        }
+        _ => {
+            unimplemented!();
+        }
+    };
+    Ok(macros_tokens)
+}
+
+fn rewrite_mod(
+    export_path: &syn::Path,
+    mut mod_path: syn::Path,
+    item_mod: &syn::ItemMod,
+    macros_tokens: &mut TokenStream,
+    uses_same_level: &TokenStream,
+) -> syn::Result<()> {
+    let mut path = export_path.to_owned();
+    let macro_content = expand_mod_path_in_macro(&mod_path, quote!(#item_mod), uses_same_level);
+    path.segments.push(syn::PathSegment {
+        ident: item_mod.ident.clone(),
+        arguments: syn::PathArguments::None,
+    });
+
+    mod_path.segments.push(syn::PathSegment {
+        ident: item_mod.ident.clone(),
+        arguments: syn::PathArguments::None,
+    });
+
+    let mut uses = TokenStream::new();
+
+    for item in item_mod.content.as_ref().unwrap().1.iter() {
+        let mod_path = mod_path.to_owned();
+        match item {
+            syn::Item::Fn(item_fn) => {
+                rewrite_fn(&path, mod_path, item_fn, macros_tokens, &uses);
+            }
+            syn::Item::Impl(item_impl) => {
+                rewrite_impl(&path, mod_path, item_impl, macros_tokens, &uses)?;
+            }
+            syn::Item::Mod(item_mod) => {
+                rewrite_mod(&path, mod_path, item_mod, macros_tokens, &uses)?;
+            }
+            syn::Item::Struct(item_struct) => {
+                rewrite_struct(item_struct, mod_path, macros_tokens, &uses);
+            }
+            syn::Item::Use(item_use) => {
+                uses.extend(quote!(#item_use));
```

Comment:
Why exactly is this stateful and order-dependent? Why do we collect `use`s and only add the ones we have seen so far into subsequent `impl`s, `mod`s, etc?

Reply:
I think it's because how rust deal with `uses`, we can have duplicate `use` in different levels but not the same level:
```rust
use prusti_contracts::*;
mod any {
  //this is valid
  use prusti_contracts::*;

  //this is error
  use prusti_contracts::*;
}

```

Therefore the `uses` is different in each level so that I think we will have to retain that structure to avoid potential conflicts, so I intend to collect `uses` at one level and pass that to the children (they will need it like they need the `mod_path` for later populating the macro)